utils/s2n_key_material.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "utils/s2n_key_material.h"
+
+#include "tls/s2n_connection.h"
+
+S2N_RESULT s2n_key_material_read(struct s2n_key_material *key_material, s2n_key_material_type type, struct s2n_blob *out)
+{
+    RESULT_ENSURE_REF(key_material);
+    RESULT_ENSURE_REF(out);
+
+    switch (type) {
+        case S2N_KEY_MATERIAL_CLIENT_MAC:
+            *out = key_material->client_mac;
+            break;
+        case S2N_KEY_MATERIAL_SERVER_MAC:
+            *out = key_material->server_mac;
+            break;
+        case S2N_KEY_MATERIAL_CLIENT_KEY:
+            *out = key_material->client_key;
+            break;
+        case S2N_KEY_MATERIAL_SERVER_KEY:
+            *out = key_material->server_key;
+            break;
+        case S2N_KEY_MATERIAL_CLIENT_IV:
+            *out = key_material->client_iv;
+            break;
+        case S2N_KEY_MATERIAL_SERVER_IV:
+            *out = key_material->server_iv;
+            break;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_key_material_update_data_offsets(struct s2n_key_material *key_material)
+{
+    RESULT_ENSURE_REF(key_material);
+
+    struct s2n_stuffer key_material_stuffer = { 0 };
+    struct s2n_blob key_material_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&key_material_blob, key_material->data, s2n_array_len(key_material->data)));
```

Comment:
Just sizeof(key_material->data), not s2n_array_len. You care about total memory, not number of elements.

Also, the `key_material->data` could cause issues since at a glance it looks like key_material is a blob. I'd give `data` a better name.